flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1433,35 +1423,16 @@ protected void init() {}
 
 		@Override
 		protected void processInput(MailboxDefaultAction.Controller controller) throws Exception {
-			holder = new LockHolder(getCheckpointLock(), latch);
-			holder.start();
-			latch.await();
-
-			// we are at the point where cancelling can happen
-			syncLatch.trigger();
-
-			// just put this to sleep until it is interrupted
-			try {
-				Thread.sleep(100000000);
-			} catch (InterruptedException ignored) {
-				// restore interruption state
-				Thread.currentThread().interrupt();
+			syncLatch.trigger(); // signal that the task can be cancelled now
+			while (task.getExecutionState() == ExecutionState.RUNNING) { // wait for the containing task to be terminated from the outside
+				try {
+					Thread.sleep(50);
+				} catch (InterruptedException e) {
+					LOG.debug("interrupted while waiting for state transition", e);
+					Thread.currentThread().interrupt();
+				}
```

Comment:
Yes, the previous version held the checkpoint lock in a separate thread while the main thread was closing the Task. But closing the task involves stopping mailbox loop, which now requires a lock (to process "poison pill").

Reply:
Is this test still testing what it was intended to test? I'm especially worried about the removed `cancelTask` method.

Reply:
To my understanding, it **was** testing that task **can** be canceled even if some thread holds the `checkpointLock`.

As now it's not possible (I think it shouldn't - please correct me if I'm wrong) the test now only checks that `task.cancelExecution` leads to shutdown when the lock is not being held.

Reply:
Addressed by not acquiring the lock for *internal* `MailboxProcessor` mails.
I.e., `allActionsCompleted()` and 3 other similar actions don't use `MailboxExecutor` and `StreamTaskActionExecutor` and go directly into `Mailbox`.

Please see hotfix commit 63d104b52a7fbca8a4d2dd2ce0f4a8d4bdad8178.

Reply:
What was the scenario that was causing the problems and motivated this change? Mailbox was stuck waiting (uninterruptibly?) on checkpoint lock acquisition by some mail, while `LockHolder` thread was blocking it? Why didn't the `TaskInterrupter` or `TaskCanceler` work?

Reply:
`TaskCanceller` thread started from `Task` was canceling `StreamTask`, which in the end was sending a mail to stop `MailboxProcessor` loop. 
But after synchronizing each mail execution, `MailboxProcessor` was not able to execute that mail - the lock was held by a thread simulating an "ill-behaving" operator from the test.
That thread wasn't interrupted, because interrupt is sent after the loop breaks.

So I removed lock acquisition for this and 3 other "internal" mails (as it reflects the original logic and doesn't have performance issues, like adding `volatile` to `mailboxLoopRunning`).
To do this I sent those mails directly from `MailboxProcessor` instead of using MailboxExecutor#executeFirst (which I removed, as it was only used for this).

Probably, we should review the order of tear-down actions in StreamTask to make it more resilient to such scenarios.

Reply:
Does this solve all of the issues What about a case when mailbox loop is stuck trying to process some normal `Mail` (fired timer), that can not be executed because of `LockHolder`?

Can we do anything in that scenario (except of relaying on the `TaskCancelerWatchDog` to shutdown whole JVM)? Wasn't equivalent of this scenario broken already before the mailbox? Isn't this the same, as pre mailbox `Task` thread blocked on `checkpointLock` acquisition in for example `StreamOneInputProcessor`?

Maybe this test is just invalid and could be dropped?

Reply:
To summarize offline discussion:

> Does this solve all of the issues?

No, it only solves this particular issue.
There are scenarios where shutting down can hang out because of mis-behaving source threads. Before the mailbox model it wasn't an issue.
To address it we need to change the shutdown behaviour (`Task` and `StreamTask`).

> Maybe this test is just invalid and could be dropped?

The test is valid but probably test for an esoteric situation.

Still, we decided to keep it and the associated change, because:
1) internal mailbox actions shouldn't acquire the lock
2) it's not trivial to enqueue an email that touches task state and doesn't acquire the lock